src/program/lwaftr/tests/propbased/genyang.lua

```diff
@@ -267,195 +274,201 @@ local function value_from_type(a_type)
-            table.insert(ids, id)
-         end
+   local function body_generator(productions)
+      local order = {}
+      local gens = {}
+      for k,v in pairs(productions) do
+         table.insert(order, k)
+         gens[k] = visit1(k, v)
+         if not v.mandatory then gens[k] = maybe(gens[k]) end
       end
-
-      local id = choose(ids)
-      if id then
-         visit(node.body[id])
-      else
-         gen_info = { node = node }
+      table.sort(order)
+      return function()
+         local ret = {}
+         for _,k in ipairs(order) do
+            local v = gens[k]()
+            if v ~= nil then table.insert(ret, v) end
+         end
+         return table.concat(ret, ' ')
       end
    end
-   function handlers.container(node)
-      path = path .. "/" .. node.id
-
-      -- don't always go into containers, since we need to test
-      -- fetching all sub-items too
-      if math.random() < 0.9 then
-         visit_body(node)
-      else
-         gen_info = { node = node }
+   function handlers.struct(keyword, production)
+      local gen = body_generator(production.members)
+      local prefix, suffix = '', ''
+      if keyword then prefix, suffix = keyword..' {', '}' end
+      return function()
+         return table.concat({prefix, gen(), suffix}, " ")
       end
    end
-   handlers['leaf-list'] = function(node)
-      if math.random() < 0.7 then
-         local idx      = choose_nat()
-         local selector = string.format("[position()=%d]", idx)
-         path = path .. "/" .. node.id .. selector
-         gen_info = { node = node, selector = idx }
-      -- sometimes omit the selector, for the benefit of commands
-      -- like add where a selector is not useful
-      else
-         path = path .. "/" .. node.id
-         gen_info = { node = node }
+   function handlers.array(keyword, production)
+      local gen = value_generator(production.element_type, generate_invalid)
+      local prefix, suffix = '', ';'
+      if keyword then prefix = keyword..' '..prefix end
+      return function()
+         local ret = {}
+         while math.random() < 0.9 do
+            table.insert(ret, prefix..gen()..suffix)
+         end
+         return table.concat(ret, " ")
       end
    end
-   function handlers.list(node)
-      local key_types = {}
-      local r = math.random()
-
-      path = path .. "/" .. node.id
-
-      -- occasionally drop the selectors
-      if r < 0.7 then
-         for key in (node.key):split(" +") do
-            key_types[key] =  node.body[key].type
-         end
-
-         for key, type in pairs(key_types) do
-            local val = assert(value_from_type(type), type.primitive_type)
-            path = path .. string.format("[%s=%s]", key, val)
-         end
-
-         -- continue path for child nodes
-         if math.random() < 0.5 then
-            visit_body(node)
-         else
-            gen_info = { node = node, selector = key_types }
+   local function shallow_copy(t)
```

Comment:
Does it make sense to move this to somewhere like `lib/yang/util.lua`? I note it's repeated here and in `schema.lua`

Reply:
Aaaah i hesitate, it's such a silly little function... in some ways i prefer to duplicate the function and not have to remember which library to import it from and document that interface etc, if it's so short...  OK if I leave this?

Reply:
Yeah it's fine. I had a feeling you didn't want it merged (I think it's come up before) but I couldn't remember the reason and I have a slight preference for it not to be duplicated but I don't mind too much either way. It can be left as is if you'd like :)

Reply:
thank you for humoring my irrationalities, i will try to keep them in check in future :)